MSAL/src/requests/MSALAccountsProvider.m

```diff
@@ -0,0 +1,155 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#import "MSALAccountsProvider.h"
+#import "MSIDDefaultTokenCacheAccessor.h"
+#import "MSALAuthority.h"
+#import "MSALAuthority_Internal.h"
+#import "MSIDAuthority.h"
+#import "MSALAccount+Internal.h"
+#import "MSIDAADNetworkConfiguration.h"
+
+@interface MSALAccountsProvider()
+
+@property (nullable, nonatomic) MSIDDefaultTokenCacheAccessor *tokenCache;
+@property (nullable, nonatomic) NSString *clientId;
+
+@end
+
+@implementation MSALAccountsProvider
+
+#pragma mark - Init
+
+- (instancetype)initWithTokenCache:(MSIDDefaultTokenCacheAccessor *)tokenCache
+                          clientId:(NSString *)clientId
+{
+    self = [super init];
+
+    if (self)
+    {
+        _tokenCache = tokenCache;
+        _clientId = clientId;
+    }
+
+    return self;
+}
+
+#pragma mark - Accounts
+
+- (void)allAccountsFilteredByAuthority:(MSALAuthority *)authority
+                       completionBlock:(MSALAccountsCompletionBlock)completionBlock;
+{
+    [authority.msidAuthority resolveAndValidate:NO
+                              userPrincipalName:nil
+                                        context:nil
+                                completionBlock:^(NSURL * _Nullable openIdConfigurationEndpoint, BOOL validated, NSError * _Nullable error) {
+                                    
+                                    if (error)
+                                    {
+                                        completionBlock(nil, error);
+                                        return;
+                                    }
+                                    
+                                    NSError *accountsError = nil;
+                                    NSArray *accounts = [self allAccountsForEnvironment:authority.msidAuthority.environment error:&accountsError];
+                                    completionBlock(accounts, accountsError);
+                                }];
+}
+
+#pragma mark - Accounts sync
+
+- (NSArray <MSALAccount *> *)allAccounts:(NSError * __autoreleasing *)error
+{
+    return [self allAccountsForEnvironment:nil error:error];
+}
+
+- (MSALAccount *)accountForHomeAccountId:(NSString *)homeAccountId
+                                   error:(NSError * __autoreleasing *)error
+{
+    NSArray<MSALAccount *> *accounts = [self allAccounts:error];
+
+    for (MSALAccount *account in accounts)
+    {
+        if ([account.homeAccountId.identifier isEqualToString:homeAccountId])
+        {
+            return account;
+        }
+    }
+
+    return nil;
+}
+
+- (MSALAccount *)accountForUsername:(NSString *)username
+                              error:(NSError * __autoreleasing *)error
+{
+    NSArray<MSALAccount *> *accounts = [self allAccounts:error];
```

Comment:
small optimization: can accountForUsername and acccountForHomeAccountId call into cache directly and match by MSIDAccount? That way we don't need to convert all MSIDAccounts into MSALAccounts.

Reply:
done